hpx/parallel/algorithms/is_sorted.hpp

```diff
@@ -0,0 +1,449 @@
+#define HPX_PARALLEL_ALGORITHMS_IS_SORTED_FEB_9_2015_0331PM
+
+#include <hpx/parallel/config/inline_namespace.hpp>
+#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/cancellation_token.hpp>
+#include <hpx/parallel/execution_policy.hpp>
+#include <hpx/parallel/util/partitioner.hpp>
+#include <hpx/parallel/util/loop.hpp>
+
+#include <algorithm>
+#include <iterator>
+#include <functional>
+
+#include <boost/static_assert.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <boost/type_traits/is_base_of.hpp>
+#include <boost/type_traits/is_same.hpp>
+
+namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
+{
+    ////////////////////////////////////////////////////////////////////////////
+    // is_sorted
+    namespace detail
+    {
+        /// \cond NOINTERNAL
+        template <typename FwdIter>
+        struct is_sorted: public detail::algorithm<is_sorted<FwdIter>, bool>
+        {
+            is_sorted()
+                : is_sorted::algorithm("is_sorted")
+            {}
+
+            template<typename ExPolicy, typename Pred>
+            static bool
+            sequential(ExPolicy const&,FwdIter first, FwdIter last,
+                Pred && pred)
+            {
+                return std::is_sorted(first, last, std::forward<Pred>(pred));
+            }
+
+            template <typename ExPolicy, typename Pred>
+            static typename detail::algorithm_result<ExPolicy, bool>::type
+            parallel(ExPolicy const& policy, FwdIter first, FwdIter last,
+                Pred && pred)
+            {
+                typedef typename std::iterator_traits<FwdIter>::reference
+                    reference;
+                typedef typename std::iterator_traits<FwdIter>::difference_type
+                    difference_type;
+                typedef typename detail::algorithm_result<ExPolicy, bool>
+                    result;
+
+                difference_type count = std::distance(first, last);
+                if (count <= 1)
+                    return result::get(true);
+
+                util::cancellation_token<> tok;
+                return util::partitioner<ExPolicy, bool>::call(
+                    policy, first, count ,
+                    [tok, pred, last](FwdIter part_begin,
+                    std::size_t part_size) mutable -> bool
+                    {
+                        FwdIter trail = part_begin++;
+                        util::loop_n(part_begin, part_size - 1,
+                            [&trail, &tok, &pred](FwdIter it)
+                            {
+                                if (pred(*it, *trail++))
+                                {
+                                    tok.cancel();
+                                }
+                            });
+                        FwdIter i = trail++;
+                        //trail now points one past the current grouping
```

Comment:
The definition of the variable `i` could be moved inside the block of the `if()` below.